packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,893 @@
+    if (finalizer != null) {
+      _finalizers.add(finalizer);
+    }
+    if (!_postFrameScheduled) {
+      _postFrameScheduled = true;
+      SchedulerBinding.instance.addPostFrameCallback((Duration _) => _doProcessing());
+    }
+  }
+
+  // TODO(goderbauer): Add API to request out-of-band serialization when no frame is scheduled.
+
+  void _doProcessing() {
+    assert(() {
+      _debugDoingUpdate = true;
+      return true;
+    }());
+    _postFrameScheduled = false;
+
+    for (final VoidCallback finalizer in _finalizers) {
+      finalizer();
+    }
+    _finalizers.clear();
+    sendToEngine(_rootBucket._rawData);
+
+    assert(() {
+      _debugDoingUpdate = false;
+      return true;
+    }());
+  }
+}
+
+/// Identifies a piece of data within or a child of a [RestorationBucket].
+///
+/// The [value] of a [RestorationId]s must be unique within a given
+/// [RestorationBucket]. [RestorationId] objects with the same [value] are
+/// considered equal.
+@immutable
+class RestorationId {
+  /// Creates a [RestorationId].
+  ///
+  /// The provided `value` must not be null.
+  const RestorationId(this.value) : assert(value != null);
+
+  /// The value of the [RestorationId].
+  ///
+  /// Two [RestorationId] objects with the same [value] are considered equal.
+  final String value;
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    return other is RestorationId && other.value == value;
+  }
+
+  @override
+  int get hashCode => hashValues(runtimeType, value);
+
+  @override
+  String toString() {
+    return '${objectRuntimeType(this, 'RestorationId')}($value)';
+  }
+}
+
+/// A [RestorationBucket] holds pieces of the restoration data that a part of
+/// the application needs to restore its state.
+///
+/// For a general overview of how state restoration works in Flutter, see the
+/// [RestorationManager].
+///
+/// [RestorationBucket]s are organized in a tree, that is rooted in
+/// [RestorationManager.rootBucket] and managed by a [RestorationManager]. The
+/// tree is serializable and must contain all the data an application needs to
+/// restore its current state at a later point in time.
+///
+/// A [RestorationBucket] stores restoration data as key-value pairs. The key is
+/// a [RestorationId] that identifies a piece of data uniquely within a bucket.
+/// The value can be anything that is serializable via the
+/// [StandardMessageCodec]. Furthermore, a [RestorationBucket] may have child
+/// buckets, which are identified within their parent via a unique
+/// [RestorationId] as well.
+///
+/// During state restoration, the data previously stored in the
+/// [RestorationBucket] hierarchy will be made available again to the
+/// application to restore it to the state it had when the data was collected.
+/// State restoration to a previous state may happen when the app is launched
+/// (e.g. after it has been terminated while running in the background) or after
```

Comment:
The thought occurred to me when reading some other docs - is it worth specifying that the type of termination we're talking about is when the OS terminates us with a warning as opposed to a hard crash (e.g. a segfault) or power failure?

Reply:
I added the word "gracefully" here. I don't have a bullet-proved definition of when OSes do and don't do state serialization, so I'd prefer to be not too specific here.